### DISCUSSION TITLE
wfi 命令について (TOPPERS/ASP + RISC-V)

### DISCUSSION THREAD

**Discussion by KAkihiroo** [toppers/users-forum#87] [Q&A]
ご担当者様

お世話になっております。
島津製作所の岸本と申します。

お忙しいところ、申し訳ございませんがご教示お願いいたします。

【背景】
現在、RISCVのアーキテクチャにて、TOPPERS/ASP Kernelを動作させようとして
いるのですが、
全てのタスクが「待ち状態」に遷移すると、割込みハンドラが起動しなくなる
症状を確認致しました。
調査した結果、ディスパッチャ（prc_support.S　dispatcher_2:）から抜けてこなくなり
（割込みハンドラが起動しない）タスク遷移が行われない状態となっています。
この現象は サンプルプログラムでも発生いたしました。

【質問】
そこで　wfi命令　について質問させてください。
この命令は、次の割込みが来るまで処理を停止するといった認識なのですが、
現在の動作では割込みを待っている感じがしません。
この命令を正常に動作させるには何か条件がございますでしょうか。

試しに　wfi命令の後にNOPを入れると、割込みが入って来るようになり、
タスク遷移出来ることは確認できております。

使用しているカーネル
asp_arch_riscv_gcc-1.9.4.tar.gz
asp-1.9.3.tar.gz

以上となります。よろしくお願いいたします。

**Comment by Ryohsuke-Takeuchi** [comment]
wfiは割込み待ちの命令ですが、NOPとは違い、命令が停止するので省エネになります。
RISC/Vに関しては、wfi命令を割込み待ちとして使えるCOREと使えないコアがあったと思います。
(2,3年前なので良く覚えていません。ARMとは違い、CPUベンダーで実装が異なります)
freeRTOSの実装では、wfi命令を使用していません。問題の発生するRISC/Vコアは、何を使用していますか？
問題が発生しているコアを、教えて頂ければ、可能ならば確認しますが、よろしくお願いいたします。

**Reply by KAkihiroo** [reply]
竹内様

ご返信ありがとうございます。
コアによっては、WFI命令が割り込み待ちとして使えないとのことで承知しました。

また、ご確認いただけるとのことで、ありがとうございます。
使用している CPU はEfinix Sapphire SoC です。
VexRiscvコアベースのSoCです。

参考までにEfinix SoC と VexRiscvコアのリポジトリのURLを添付いたします。
- [Efinix SoC](https://www.efinixinc.com/jp/products-riscv.html)
- [VexRiscv](https://github.com/SpinalHDL/VexRiscvSoftcoreContest2018)

お手数をおかけしますがよろしくお願いいたします。

**Comment by Ryohsuke-Takeuchi** [comment]
RISC-V用のprc_support.SをTOPPERS_CUSTOM_IDLE設定で、変更できるものを今年中にリリースします。
VexRiscvについては、xyloniを買って、sapphire SoCを生成して書き込んで、
EFINITY RISC-V Embedded Software IDE 2022.2でデバッグする環境を作っていますが、
JTAGデバッグでエラーが発生して、先に進みません。xyloni-master中のサンプルIPのtop_opalSoc.hexを書き込めば、JTAGデバッグできますが、FPGAを生成できないとRAMサイズ4KBではRTOSは実行できないので、まだまだ、時間がかかりそうなのです。
xyloniでは、メモリサイズで足りなくでダメですし、sapphire SoCを生成できないとRAMサイズも変更できませんし。
よろしくお願いいたします。

**Reply by KAkihiroo** [reply]
ご対応いただきありがとうございます。
大変助かります。

弊社では、Trion T120 BGA324 評価ボードでsapphire SoC を導入し、TOPPERSサンプルプログラムをポーティングして動作確認を行っておりました。
その時のソースコードやプロジェクトが参考になるかと思いますので、共有できないか社内で相談し、もし可能なら共有いたします。

よろしくお願いいたします。

**Comment by Ryohsuke-Takeuchi** [comment]
岸本様

竹内です。
xyloniボードでデバッグできるようになったので、受信割込みのシリアルドライバーのテストプログラムを作成し、受信割込みで受け取ったデータを取り出す関数ReciveChar()の前に"WFI"命令を入れて実行させたのですが、WFI命令はNOPと同じように動作しているように思います。WFI命令の後にNOP(asm("nop":::"memory");)を入れないと割込みが発生しません。
Sapphire SoCは、最新のEfinity 2022.2で生成しました。
WFI命令を使用する意味がないのですが、割込み許可から割込み禁止までの間に２命令実行しないと割込みを受け付けないようです。
prc_support.Sの割込み待ちは以下のようにすれば問題ないようです。

`

dispatcher_2:
	/*
	 *  割込みを許可し，割込みモードに切り換えて，割込みを待つ．
	 *
	 *  ここで割込みモードに切り換えるのは，ここで発生する割込み処理に
	 *  どのスタックを使うかという問題の解決と，割込みハンドラ内でのタ
	 *  スクディスパッチの防止という2つの意味がある．
	 *
	 *  プロセッサを割込み待ちに移行させる処理と，割込み許可とは，不可
	 *  分に行なう必要がある（M68040ではstop命令で両方行なうので問題な
	 *  い）．これを不可分に行なわない場合，割込みを許可した直後に割込
	 *  みが入り，その中でタスクが実行可能状態になると，実行すべきタス
	 *  クがあるにもかかわらずプロセッサが割込み待ちになってしまう．
	 *
	 *  割込み待ちの間は，p_runtskをNULL（＝0）に設定しなければならな
	 *  い．このように設定しないと，割込みハンドラからiget_tidを呼び出
	 *  した際の動作が仕様に合致しなくなる．
	 **/
#ifdef TOPPERS_CUSTOM_IDLE
	toppers_asm_custom_idle
#else
	la a1, kernel_mie				/* 割込み待ち */
	LREG a2, 0(a1)
	csrs mie, a2
	nop
	nop
	csrw mie, a5					/* 元の状態に戻す */
#endif
	la a2, reqflg					/* reqflgがfalseならdispatcher_2へ */
	lw a1, 0(a2)
	beqz a1, dispatcher_2
	sw zero, 0(a2)					/* reqflgをfalseに */

`

以下検証したプログラム：

`////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2013-2022 Efinix Inc. All rights reserved.              
//
// This   document  contains  proprietary information  which   is        
// protected by  copyright. All rights  are reserved.  This notice       
// refers to original work by Efinix, Inc. which may be derivitive       
// of other work distributed under license of the authors.  In the       
// case of derivative work, nothing in this notice overrides the         
// original author's license agreement.  Where applicable, the           
// original license agreement is included in it's original               
// unmodified form immediately below this header.                        
//                                                                       
// WARRANTY DISCLAIMER.                                                  
//     THE  DESIGN, CODE, OR INFORMATION ARE PROVIDED “AS IS” AND        
//     EFINIX MAKES NO WARRANTIES, EXPRESS OR IMPLIED WITH               
//     RESPECT THERETO, AND EXPRESSLY DISCLAIMS ANY IMPLIED WARRANTIES,  
//     INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF          
//     MERCHANTABILITY, NON-INFRINGEMENT AND FITNESS FOR A PARTICULAR    
//     PURPOSE.  SOME STATES DO NOT ALLOW EXCLUSIONS OF AN IMPLIED       
//     WARRANTY, SO THIS DISCLAIMER MAY NOT APPLY TO LICENSEE.           
//                                                                       
// LIMITATION OF LIABILITY.                                              
//     NOTWITHSTANDING ANYTHING TO THE CONTRARY, EXCEPT FOR BODILY       
//     INJURY, EFINIX SHALL NOT BE LIABLE WITH RESPECT TO ANY SUBJECT    
//     MATTER OF THIS AGREEMENT UNDER TORT, CONTRACT, STRICT LIABILITY   
//     OR ANY OTHER LEGAL OR EQUITABLE THEORY (I) FOR ANY INDIRECT,      
//     SPECIAL, INCIDENTAL, EXEMPLARY OR CONSEQUENTIAL DAMAGES OF ANY    
//     CHARACTER INCLUDING, WITHOUT LIMITATION, DAMAGES FOR LOSS OF      
//     GOODWILL, DATA OR PROFIT, WORK STOPPAGE, OR COMPUTER FAILURE OR   
//     MALFUNCTION, OR IN ANY EVENT (II) FOR ANY AMOUNT IN EXCESS, IN    
//     THE AGGREGATE, OF THE FEE PAID BY LICENSEE TO EFINIX HEREUNDER    
//     (OR, IF THE FEE HAS BEEN WAIVED, $100), EVEN IF EFINIX SHALL HAVE 
//     BEEN INFORMED OF THE POSSIBILITY OF SUCH DAMAGES.  SOME STATES DO 
//     NOT ALLOW THE EXCLUSION OR LIMITATION OF INCIDENTAL OR            
//     CONSEQUENTIAL DAMAGES, SO THIS LIMITATION AND EXCLUSION MAY NOT   
//     APPLY TO LICENSEE.                                                
//
////////////////////////////////////////////////////////////////////////////////
#include <stdint.h>
#include "clint.h"
#include "bsp.h"
#include "riscv.h"
#include "serial.h"


extern void BeforeInit(void);
extern void sys_putc(char c);



void crash(){
    bsp_printf("** CRASH **\r\n");
    while(1);
}

void main() {
    int c;

    BeforeInit();
    InitSerial(0);

    bsp_printf("\r\ntest interrupt serial driver. \r\n");
    csr_clear(mie, MIE_MEIE);
    while(1){
        csr_set(mie, MIE_MEIE);
        asm("wfi":::"memory");
        asm("nop":::"memory");
        csr_clear(mie, MIE_MEIE);
        if((c = ReciveChar(0)) != 0){
            sys_putc(c);
            sys_putc('\n');
        }
        else{
            bsp_printf("pass !\r\n");
            sil_dly_nse(500*1000*1000);
        }
    }
}
`
mainのアセンブラは以下の通りです。

`

void main() {
f90007f8:	ff010113          	addi	sp,sp,-16
f90007fc:	00112623          	sw	ra,12(sp)
	int c;

    BeforeInit();
f9000800:	2d4000ef          	jal	ra,f9000ad4 <BeforeInit>
	InitSerial(0);
f9000804:	00000513          	li	a0,0
f9000808:	8edff0ef          	jal	ra,f90000f4 <InitSerial>

    bsp_printf("\r\ntest interrupt serial driver. \r\n");
f900080c:	f9001537          	lui	a0,0xf9001
f9000810:	c4450513          	addi	a0,a0,-956 # f9000c44 <__global_pointer$+0xfffff7dc>
f9000814:	e95ff0ef          	jal	ra,f90006a8 <bsp_printf>
	csr_clear(mie, MIE_MEIE);
f9000818:	000017b7          	lui	a5,0x1
f900081c:	80078793          	addi	a5,a5,-2048 # 800 <__stack_size+0x600>
f9000820:	3047b073          	csrc	mie,a5
f9000824:	0180006f          	j	f900083c <main+0x44>
		if((c = ReciveChar(0)) != 0){
			sys_putc(c);
			sys_putc('\n');
        }
		else{
			bsp_printf("pass !\r\n");
f9000828:	80018513          	addi	a0,gp,-2048 # f9000c68 <__init_array_end+0xa8>
f900082c:	e7dff0ef          	jal	ra,f90006a8 <bsp_printf>
			sil_dly_nse(500*1000*1000);
f9000830:	1dcd6537          	lui	a0,0x1dcd6
f9000834:	50050513          	addi	a0,a0,1280 # 1dcd6500 <__stack_size+0x1dcd6300>
f9000838:	2d8000ef          	jal	ra,f9000b10 <sil_dly_nse>
	    csr_set(mie, MIE_MEIE);
f900083c:	000017b7          	lui	a5,0x1
f9000840:	80078793          	addi	a5,a5,-2048 # 800 <__stack_size+0x600>
f9000844:	3047a073          	csrs	mie,a5
		asm("wfi":::"memory");
f9000848:	10500073          	wfi
		asm("nop":::"memory");
f900084c:	00000013          	nop
		csr_clear(mie, MIE_MEIE);
f9000850:	3047b073          	csrc	mie,a5
		if((c = ReciveChar(0)) != 0){
f9000854:	00000513          	li	a0,0
f9000858:	b01ff0ef          	jal	ra,f9000358 <ReciveChar>
f900085c:	fc0506e3          	beqz	a0,f9000828 <main+0x30>
			sys_putc(c);
f9000860:	07c000ef          	jal	ra,f90008dc <sys_putc>
			sys_putc('\n');
f9000864:	00a00513          	li	a0,10
f9000868:	074000ef          	jal	ra,f90008dc <sys_putc>
f900086c:	fd1ff06f          	j	f900083c <main+0x44>

`

![term](https://user-images.githubusercontent.com/91311953/212840706-67408560-2a2b-40dd-a0a6-565aac847506.png)

**Comment by Ryohsuke-Takeuchi** [comment]
T120 BGA324ボードを入手し、ASPカーネルを作成しました。ASPのRISC-V版の個別パッケージとしてリリースしましたので、参照してください。デバッグ方法は、target/t120bga_gcc/README.txtに記載しています。よろしくお願いいたします。

**Comment by KAkihiroo** [comment]
竹内様
返信できておらず申し訳ありません。また、検証いただきありがとうございます。
リリースいただきありがとうございます。こちらでも動作確認いたします。

> WFI命令はNOPと同じように動作しているように思います。

> 割込み許可から割込み禁止までの間に２命令実行しないと割込みを受け付けないようです。

私も同じ認識です。  
私も竹内様と同じく、割込み許可から割り込み禁止までの間にNOPを2つ挟む実装としました。  
動作も問題なさそうでしたが、NOPの数については、CPUのクロック周波数などによって決めるべきでしょうか？  
もしご意見などあれば、伺いたく思います。

よろしくお願いいたします。
